pycode/memilio-epidata/memilio/epidata/assessNPIEffects.py

```diff
@@ -0,0 +1,595 @@
+            plt.text(-0.05, y_lower + 0.5 * size_cluster_i, str(i))
+
+            y_lower = y_upper + 10
+
+        plt.xlabel("The silhouette coefficient values")
+        plt.ylabel("Cluster label")
+
+        plt.axvline(x=0, c='black', alpha=0.3)
+        plt.axvline(x=0.25, c='black', alpha=0.3)
+        plt.axvline(x=0.5, c='black', alpha=0.3)
+        plt.axvline(x=0.75, c='black', alpha=0.3)
+
+        plt.yticks([])  # Clear the yaxis labels / ticks
+        plt.xticks([-0.4, -0.2, 0, 0.2, 0.4, 0.6, 0.8, 1])
+
+        plt.title(
+            "Silhouette analysis for " + label +
+            " clustering with n_clusters = " + str(n_clusters),
+        )
+        plt.tight_layout()
+        plt.show()
+
+        return sample_silhouette_values
+
+
+def analyze_npi_data(
+        read_data, make_plot, fine_resolution, npis, directory, file_format,
+        npi_codes_considered):
+
+    if not read_data:
+        raise FileNotFoundError('')
+        # transform_npi_data(fine_resolution=2,
+        #                file_format=dd.defaultDict['file_format'],
+        #                out_folder=dd.defaultDict['out_folder'],
+        #                start_date=dd.defaultDict['start_date'],
+        #                end_date=dd.defaultDict['end_date'],
+        #                make_plot=dd.defaultDict['make_plot'],
+        #                )
+
+    else:  # read formatted file
+        npis = pd.read_excel(
+            os.path.join(
+                directory, 'datensatzbeschreibung_massnahmen.xlsx'),
+            sheet_name=3, engine='openpyxl')
+        if fine_resolution > 0:
+            if fine_resolution == 1:
+                filename = 'germany_counties_npi_subcat_incgrouped'
+            else:
+                filename = 'germany_counties_npi_subcat'
+        else:
+            filename = 'germany_counties_npi_maincat'
+        df_npis = pd.read_csv(directory + filename + ".csv")
+    try:
+        df_npis = df_npis.drop('Unnamed: 0', axis=1)
+    except KeyError:
+        pass
+    # df_npis = mdfs.extract_subframe_based_on_dates(df_npis, date(2021,1,1), date(2021,6,1))
+    npis = pd.read_json(os.path.join(directory, 'npis.json'))
+    # get code levels (main/subcodes) and position of main codes
+    # code_level = [i.count('_') for i in npi_codes]
+    # main_code_pos = [i for i in range(len(code_level)) if code_level[i] == 1]
+
+    # check if any other integer than 0: not implemented or 1: implemented is
+    # used (maybe to specify the kind of implementation)
+
+    all_subcodes = [x for x in npis.NPI_code if len(x) <=8]
+    df_merged = df_npis.iloc[:, :2]
+    for subcode in all_subcodes:
+        # extract columns which have the subcode as part of the column
+        # name and sum over all these subcodes
+        df_merged[subcode] = df_npis.filter(
+            regex=subcode).sum(axis=1)
+    #npi_codes_considered = npis.NPI_code.values.tolist()
+    npi_codes_considered = all_subcodes
+    df_npis = df_merged.copy()
```

Comment:
this is a very quick-and-dirty implementation.

Reply:
besides the naming of all_subcodes and npi_codes_considered (which I will change), this is done the same way as in getNPIData.py.